fdbserver/RestoreMaster.actor.cpp

```diff
@@ -0,0 +1,1339 @@
+
+				if (allLoadReqsSent) {
+					printf("[INFO] allLoadReqsSent has finished.\n");
+					break; // NOTE: need to change when change to wait on any cmdReplies
+				}
+
+			} catch (Error &e) {
+				fprintf(stdout, "[ERROR] Node:%s, Commands before cmdID:%s error. error code:%d, error message:%s\n", self->describeNode().c_str(),
+						self->cmdID.toString().c_str(), e.code(), e.what());
+				curFileIndex = checkpointCurFileIndex;
+				curOffset = checkpointCurOffset;
+			}
+		}
+
+		if (phaseType == RestoreCommandEnum::Assign_Loader_Log_File) {
+			phaseType = RestoreCommandEnum::Assign_Loader_Range_File;
+		} else if (phaseType == RestoreCommandEnum::Assign_Loader_Range_File) {
+			break;
+		}
+	}
+
+	wait( delay(1.0) );
+	printf("[Progress] distributeWorkloadPerVersionBatch loadFiles time:%.2f seconds\n", now() - startTime);
+
+	ASSERT( cmdReplies.empty() );
+	
+	wait( delay(5.0) );
+	// Notify the applier to applly mutation to DB
+
+	startTime = now();
+	wait( notifyApplierToApplyMutations(self) );
+	printf("[Progress] distributeWorkloadPerVersionBatch applyToDB time:%.2f seconds\n", now() - startTime);
+
+	state double endTime = now();
+
+	double runningTime = endTime - startTimeBeforeSampling;
+	printf("[Progress] Node:%s distributeWorkloadPerVersionBatch runningTime without sampling time:%.2f seconds, with sampling time:%.2f seconds\n",
+			self->describeNode().c_str(),
+			runningTime, endTime - startTimeAfterSampling);
+
+	return Void();
+
+}
+
+
+// RestoreMaster: Ask loaders to sample data and send mutations to master applier. Ask master applier to calculate the range for each applier
+ACTOR static Future<Void> sampleWorkload(Reference<RestoreMasterData> self, RestoreRequest request, Reference<RestoreConfig> restoreConfig, int64_t sampleMB_input) {
+	state Key tagName = request.tagName;
+	state Key url = request.url;
+	state bool waitForComplete = request.waitForComplete;
+	state Version targetVersion = request.targetVersion;
+	state bool verbose = request.verbose;
+	state KeyRange restoreRange = request.range;
+	state Key addPrefix = request.addPrefix;
+	state Key removePrefix = request.removePrefix;
+	state bool lockDB = request.lockDB;
+	state UID randomUid = request.randomUid;
+	state Key mutationLogPrefix = restoreConfig->mutationLogPrefix();
+
+	state bool allLoadReqsSent = false;
+	state int64_t sampleMB = sampleMB_input; //100;
+	state int64_t sampleB = sampleMB * 1024 * 1024; // Sample a block for every sampleB bytes. // Should adjust this value differently for simulation mode and real mode
+	state int64_t curFileIndex = 0;
+	state int64_t curFileOffset = 0;
+	state int64_t loadSizeB = 0;
+	state int64_t loadingCmdIndex = 0;
+	state int64_t sampleIndex = 0;
+	state double totalBackupSizeB = 0;
+	state double samplePercent = 0.05; // sample 1 data block per samplePercent (0.01) of data. num_sample = 1 / samplePercent
+
+	// We should sample 1% data
```

Comment:
We're going to be limited by the amount of memory that the master applier has, so this seems like we should basically have a cap of how much data we can sample in total of ... 4GB?